File: algorithm/src/services/recommendationService.js
/**
 * 1. Tries Gemini for genre selection
 * 2. Falls back to band
 * 3. Fetches candidate tracks from SoundCloud
 * 4. Filters out unplayable tracks
 * 5. Weighted-pick among remaining candidates, favoring liked artists
 */


/**One function recommendationRoutes.js calls.  */
const scoreSongs = require('../utils/scoreSongs');
const filterSongs = require('../utils/filterSongs');
const soundcloudService = require('./soundcloudService');
const geminiService = require('./geminiService');
const { swipeHistory: mockSwipeHistory } = require('../data/mockUser');

async function getRecommendation({ genres, excludeIds = [], swipeHistory, preferredGenres = [] }) {
  if (!Array.isArray(genres) || genres.length === 0) {
    return { status: 'error', message: 'genres array is required' };
  }
 
  // No history supplied (e.g. manual testing against this service directly)
  // falls back to a fixture so the algorithm still has something to react to.
  const history = Array.isArray(swipeHistory) && swipeHistory.length
    ? swipeHistory
    : mockSwipeHistory;
 
  let genre = null;
  let usedGemini = false;
 
  if (geminiService.isEnabled()) {
    try {
      const suggestion = await geminiService.suggestGenre(history, genres);
      if (suggestion) {
        genre = suggestion;
        usedGemini = true;
      }
    } catch (err) {
      console.warn(`[recommendationService] Gemini suggestion failed, falling back to bandit: ${err.message}`);
    }
  }
 
  if (!genre) {
    genre = scoreSongs.pickGenre(history, genres, preferredGenres);
  }
 
  const rawCandidates = await soundcloudService.getTracksByGenre(genre);
  const candidates = filterSongs(rawCandidates, excludeIds);
 
  if (!candidates.length) {
    return { status: 'error', message: `no candidates found for genre: ${genre}` };
  }
 
  const song = scoreSongs.chooseTrack(candidates, history);
  const score = scoreSongs.confidenceScore(genre, song, history);
 
  return { status: 'ok', song, chosenGenre: genre, usedGemini, score };
}
 
module.exports = { getRecommendation };

File: algorithm/src/services/recommendationService.py
"""
 * 1. Tries Gemini for genre selection
 * 2. Falls back to band
 * 3. Fetches candidate tracks from SoundCloud
 * 4. Filters out unplayable tracks
 * 5. Weighted-pick among remaining candidates, favoring liked artists

One function recommendationRoutes.js calls.
"""

from src.utils import score_songs
from src.utils.filter_songs import filter_songs
from src.services import soundcloud_service
from src.services import gemini_service
from src.data.mock_user import SWIPE_HISTORY as MOCK_SWIPE_HISTORY

def get_recommendation(genres=None, exclude_ids=None, swipe_history=None, preferred_genres=None):
    genres = genres or []
    exclude_ids = exclude_ids or []
    preferred_genres = preferred_genres or []
 
    if not isinstance(genres, list) or len(genres) == 0:
        return {'status': 'error', 'message': 'genres array is required'}
    history = swipe_history if (isinstance(swipe_history, list) and len(swipe_history) > 0) else MOCK_SWIPE_HISTORY
 
    genre = None
    used_gemini = False
 
    if gemini_service.is_enabled():
        try:
            suggestion = gemini_service.suggest_genre(history, genres)
            if suggestion:
                genre = suggestion
                used_gemini = True
        except Exception as e:
            print(f'[recommendation_service] Gemini suggestion failed, falling back to bandit: {e}')
 
    if not genre:
        genre = score_songs.pick_genre(history, genres, preferred_genres)
 
    raw_candidates = soundcloud_service.get_tracks_by_genre(genre)
    candidates = filter_songs(raw_candidates, exclude_ids)
 
    if not candidates:
        return {'status': 'error', 'message': f'no candidates found for genre: {genre}'}
 
    song = score_songs.choose_track(candidates, history)
    score = score_songs.confidence_score(genre, song, history)
 
    return {'status': 'ok', 'song': song, 'chosenGenre': genre, 'usedGemini': used_gemini, 'score': score}